### PR TITLE
rgw: add tenant support to multisite sync

### PR DESCRIPTION
**Pull request by cbodley**
Initial support for syncing buckets/objects created by tenant users. Needs more testing and cleanup.

Fixes: http://tracker.ceph.com/issues/16469

### CODE REVIEW

File: src/rgw/rgw_bucket.cc [hunk 1]
```diff
@@ -295,29 +295,28 @@ int rgw_bucket_instance_remove_entry(RGWRados *store, string& entry, RGWObjVersi
   return store->meta_mgr->remove_entry(bucket_instance_meta_handler, entry, objv_tracker);
 }
 
+// parse instance in format: [tenant:]instance[:shard_id], returning
```
Comment (yehudasa) [inline src/rgw/rgw_bucket.cc]: @cbodley this is ambiguous. can be either tenant:instance, or instance:shard_id. Will only work if we're guaranteed that instance does not parse as integer (and shard_id does). Do we want to tie our hands like that?

File: src/rgw/rgw_bucket.cc [hunk 2]
```diff
@@ -319,6 +321,53 @@ int rgw_bucket_parse_bucket_instance(const string& bucket_instance, string *targ
...
 }
 
+// parse key in format: [tenant/]name:instance[:shard_id]
+int rgw_bucket_parse_bucket_key(CephContext *cct, const string& key,
+                                rgw_bucket *bucket, int *shard_id)
+{
+  boost::string_ref name{key};
+  boost::string_ref instance;
+
+  // split tenant:name
```
Comment (yehudasa) [inline src/rgw/rgw_bucket.cc]: comment should be corrected? tenant/name

File: src/rgw/rgw_bucket.cc (hunk 1 above)
Comment (cbodley) [inline src/rgw/rgw_bucket.cc]: is it safe to assume that all bucket instances will have the format "name:zone.instance_id.bucket_id" that results from this line in `RGWRados::create_bucket()`?
```c++
snprintf(buf, sizeof(buf), "%s.%llu.%llu", get_zone_params().get_id().c_str(), (long long)iid, (long long)bid);
```
if so, we can resolve the ambiguity by parsing ':' as shard_id only if it comes after the '.'s

Comment (yehudasa) [inline src/rgw/rgw_bucket.cc]: actually it's not even safe to assume that all buckets will even have a bucket instance (e.g., buckets created pre dumpling). I'm not sure whether it's going to always have the same format (could be that the zone was added later). We really shouldn't rely on that string format.

**Comment by cbodley** [comment]
@yehudasa I pushed an updated patch set that changes the bucket instance metadata keys to "tenant/" so we don't have to worry about parsing "tenant:" vs. ":shard" in the datalog

currently passes test_multi.py with all 4 combinations of tenant vs. non-tenant user and rgw_override_bucket_index_max_shards=0 vs. 2, so parsing of bucket instance keys is working in all cases

i'm starting to look into testing the backward compatibility

**Comment by cbodley** [comment]
with the addition of the patch "rgw: convert bucket instance listings back to metadata key format", I was able to get preexisting tenant's buckets to sync after running `radosgw-admin data sync init`

**Comment by cbodley** [comment]
also pushed to https://github.com/ceph/ceph/tree/wip-rgw-tenant-sync so i can schedule a teuthology run tomorrow

**Comment by cbodley** [comment]
http://pulpito.ceph.com/cbodley-2016-07-07_11:21:33-rgw-wip-rgw-tenant-sync---basic-mira/ has 6 failed and 3 dead tests. one was a valgrind issue in the osd, and the rest appear to be networking issues (failure to download packages, timeouts, etc)

**Comment by cbodley** [comment]
added a patch to skip datalog entries for missing buckets, so that old entries missing the 'tenant/' won't stay in the error_repo forever

**Comment by yehudasa** [comment]
@cbodley lgtm, should merge if passing tests (and get it backported to jewel)

**Pull request merged**